### PR TITLE
PERF: use StringHasTable for strings

### PR DESCRIPTION
**Pull request by jreback**
xref #13745 

provides a modest speedup for all string hashing. The key thing is, it will release the GIL on more operations where this is possible (mainly ``factorize``).

can be easily extended to ``.value_counts()`` and ``.duplicated()`` (for strings), new issue for that one

```
In [9]: np.random.seed(1234)

In [10]: strings = tm.makeStringIndex(1000000)

In [11]: def f():
    ...:     for i in range(2):
    ...:         pd.factorize(strings)
    ...:         

In [12]: @tm.test_parallel(num_threads=2)
    ...: def g():
    ...:     pd.factorize(strings)
    ...:     

In [13]: %timeit f()
1 loop, best of 3: 685 ms per loop

In [14]: %timeit g()
1 loop, best of 3: 446 ms per loop

In [15]: strings = strings.take(np.random.randint(0,1000,size=len(strings)))

In [16]: strings.nunique()
Out[16]: 1000

In [17]: %timeit f()
1 loop, best of 3: 222 ms per loop

In [18]: %timeit g()
10 loops, best of 3: 190 ms per loop
```

### CODE REVIEW

**Comment by jreback** [comment]
cc @mrocklin

File: asv_bench/benchmarks/algorithms.py [hunk 1]
```diff
@@ -23,12 +23,16 @@ def setup(self):
...
         self.arrneg = np.arange(-1000000, 0)
         self.arrmixed = np.array([1, -1]).repeat(500000)
+        self.strings = tm.makeStringIndex(100000)
 
         # match
         self.uniques = tm.makeStringIndex(1000).values
         self.all = self.uniques.repeat(10)
 
     def time_factorize_int(self):
+        self.strings.factorize()
```
Comment (jreback) [inline asv_bench/benchmarks/algorithms.py]: should be factorize_string

File: doc/source/whatsnew/v0.20.0.txt [hunk 2]
```diff
@@ -110,7 +110,7 @@ Removal of prior version deprecations/changes
 Performance Improvements
 ~~~~~~~~~~~~~~~~~~~~~~~~
 
-
+- increased performance of ``pd.factorize()`` by releasing the GIL with ``object`` dtype when inferred as strings (:issue:``)
```
Comment (jreback) [inline doc/source/whatsnew/v0.20.0.txt]: add number

File: pandas/src/hashtable_class_helper.pxi.in [hunk 3]
```diff
@@ -1,3 +1,5 @@
+# cython: profile=True
+
```
Comment (jreback) [inline pandas/src/hashtable_class_helper.pxi.in]: change back to False

File: pandas/src/hashtable_class_helper.pxi.in [hunk 4]
```diff
@@ -94,6 +99,61 @@ cdef class {{name}}Vector:
 
 {{endfor}}
 
+cdef class StringVector:
+
```
Comment (jreback) [inline pandas/src/hashtable_class_helper.pxi.in]: not actually using this so add comment /
delete

File: asv_bench/benchmarks/gil.py [hunk 5]
```diff
@@ -379,3 +379,44 @@ def pg_read_csv_datetime(self):
...
+    def _pg_factorize_strings_2(self):
+        self.factorize_strings()
+
+    def time_factorize_strings_2(self):
+        for i in range(4):
+            self._pg_factorize_strings_2()
+
+    def time_factorize_strings(self):
+        for i in range(8):
+            self.factorize_strings()
```
Comment (jorisvandenbossche) [inline asv_bench/benchmarks/gil.py]: Is it needed to have all the different number of threads? (most other tests in this file just use one number of threads to test). It's just that it adds, possibly unnecessarily, to the number of tests / time that it takes to run the benchmarks

**Comment by jreback** [comment]
cc @mrocklin

if you can give this a stress test (and see if it helps in your perf tests), when you have a chance

**Comment by mrocklin** [comment]
I *think* I've narrowed down my original problem to something else.
Working on isolating it further now.  Will try this after that issue is
removed.

**Pull request closed without merging**